src/software/ai/evaluation/q_learning/attacker_mdp_feature_extractor.h

```diff
@@ -0,0 +1,78 @@
+#pragma once
+
+#include "software/ai/evaluation/q_learning/attacker_mdp_action.h"
+#include "software/ai/evaluation/q_learning/feature_extractor.hpp"
+
+/**
+ * State representation of the Markov decision process (MDP) modelling
+ * the Attacker agent's gameplay decision making.
+ */
+struct AttackerMdpState
+{
+    WorldPtr world_ptr;
+    std::shared_ptr<Strategy> strategy;
+};
+
+/**
+ * FeatureExtractor that extracts the values of features from a AttackerMdpState.
+ */
+class AttackerMdpFeatureExtractor
+    : public FeatureExtractor<AttackerMdpState, AttackerMdpAction>
+{
+   public:
+    explicit AttackerMdpFeatureExtractor();
+
+    /**
+     * Extracts the value of a feature representing the ball's x-distance
+     * to the enemy goal line.
+     *
+     * @param state the state to extract the feature from
+     *
+     * @return the value of the feature
+     */
+    static double ballXPositionFeature(const AttackerMdpState& state);
+
+    /**
+     * Extracts the value of a feature representing the rating of the current
+     * best pass the Attacker agent can take.
+     *
+     * @param state the state to extract the feature from
+     *
+     * @return the value of the feature
+     */
+    static double bestPassRatingFeature(const AttackerMdpState& state);
+
+    /**
+     * Extracts the value of a feature representing the open angle of the current
+     * best shot on goal the Attacker agent can take.
+     *
+     * @param state the state to extract the feature from
+     *
+     * @return the value of the feature
+     */
+    static double bestShotOpenAngleFeature(const AttackerMdpState& state);
+
+    /**
+     * Extracts the value of a feature representing the number of friendly
+     * robots on the field. The fewer robots on the field, the larger the 
+     * feature value (i.e. this feature is more important when we are down
+     * robots).
+     * 
+     * @param state the state to extract the feature from
+     *
+     * @return the value of the feature
+     */
+    static double numFriendlyRobotsFeature(const AttackerMdpState& state);
+
+    /**
+     * Extracts the value of a feature representing the number of enemy
+     * robots on the field. The fewer robots on the field, the larger the 
+     * feature value (i.e. this feature is more important when they are down
+     * robots).
```

Comment:
the implementation doesn't quite match the docs. Should it be `1-normalizeValueToRange....`

Reply:
I've set the output range for `normalizeValueToRange` to go from 1 to 0, so the output starts at 1 for `DIV_B_NUM_ROBOTS` and it approaches 0 as `num_robots` gets smaller

Reply:
am I missing something, it seems like the implementation for `numFriendlyRobotsFeature` and `numEnemyRobotFeature` normalizes it the same way. Shouldn't one favour more robots (friendly)? Let me know if I'm understanding it completely incorrectly.

Reply:
The weight $`w`$ associated with `numFriendlyRobotsFeature` and a given action indicates how important `numFriendlyRobotsFeature` is to the success/viability of that action. So as the number of robots decreases, `numFriendlyRobotsFeature` increases and the $`w`$ * `numFriendlyRobotsFeature` term will contribute more to the Q-value for that action. This makes certain actions more likely to be selected when we have less robots. 

Same logic goes for `numEnemyRobotFeature`

If a feature has a value of 0, that feature won't contribute to the Q-value. So I set `numFriendlyRobotsFeature` and `numEnemyRobotFeature` to 0 when the teams have the max number of robots to signify that "the number of robots is not important to us right now."

Reply:
Ah I see that makes sense